core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala

```diff
@@ -34,22 +34,34 @@ import scala.collection.immutable.SortedSet
  * A [[Consumer]] that fetches all messages available in a topic up to the last message
  * at the time it starts running.
  */
-class FullConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
+class SeekConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
+    initialOffsets: Offsets,
     settings: ConsumerSettings[F, K, V],
     decoder: RecordDecoder[F, K, V])
     extends Consumer[F] with Logging {
 
-  import FullConsumer.topicPartitionOrder
+  import SeekConsumer.topicPartitionOrder
 
   val F: ConcurrentEffect[F] = ConcurrentEffect[F]
 
-  override def fetch(topic: String): Resource[F, Stream[F, Byte]] = {
+  override def fetch(topic: String): Resource[F, (Offsets, Stream[F, Byte])] = {
     consumerResource[F].using(settings) evalMap { consumer =>
       for {
         endOffsets <- assignNonEmptyPartitionsForTopic(consumer, topic)
```

Comment:
If we separate obtaining non-empty partitions from assigning them to the consumer, we could detect which ones actually have new records and only consume from them, allowing us to avoid the off-by-one seek and the `filter`.